apps/live/src/hocuspocus.ts

```diff
@@ -0,0 +1,200 @@
+   * @param token - The token
+   * @returns The authenticated user
+   */
+  private onAuthenticate = async ({ requestHeaders, context, token }: any) => {
+    let cookie: string | undefined = undefined;
+    let userId: string | undefined = undefined;
+
+    // Extract cookie (fallback to request headers) and userId from token (for scenarios where
+    // the cookies are not passed in the request headers)
+    try {
+      const parsedToken = JSON.parse(token) as TUserDetails;
+      userId = parsedToken.id;
+      cookie = parsedToken.cookie;
+    } catch (error) {
+      // If token parsing fails, fallback to request headers
+      console.error("Token parsing failed, using request headers:", error);
+    } finally {
+      // If cookie is still not found, fallback to request headers
+      if (!cookie) {
+        cookie = requestHeaders.cookie?.toString();
+      }
+    }
+
+    if (!cookie || !userId) {
+      throw new Error("Credentials not provided");
+    }
+
+    // set cookie in context, so it can be used throughout the ws connection
+    (context as HocusPocusServerContext).cookie = cookie;
+
+    try {
+      await handleAuthentication({
+        cookie,
+        userId,
+      });
+    } catch (_error) {
+      throw Error("Authentication unsuccessful!");
+    }
+  };
+
+  private onStateless = async ({ payload, document }: any) => {
+    // broadcast the client event (derived from the server event) to all the clients so that they can update their state
+    const response = DocumentCollaborativeEvents[payload as TDocumentEventsServer].client;
+    if (response) {
+      document.broadcastStateless(response);
+    }
+  };
+
+  private onDatabaseFetch = async ({ context, documentName: pageId, requestParameters }: any) => {
+    const cookie = (context as HocusPocusServerContext).cookie;
+    // query params
+    const params = requestParameters;
+    const documentType = params.get("documentType")?.toString() as TDocumentTypes | undefined;
+    return new Promise(async (resolve) => {
+      try {
+        let fetchedData = null;
+        if (documentType === "project_page") {
+          fetchedData = await fetchPageDescriptionBinary(params, pageId, cookie);
+        } else {
+          fetchedData = await fetchDocument({
+            cookie,
+            documentType,
+            pageId,
+            params,
+          });
+        }
+        resolve(fetchedData);
+      } catch (error) {
+        logger.error("Error in fetching document", error);
+      }
+    });
+  };
+
+  private onDatabaseStore = async ({ context, state, documentName: pageId, requestParameters }: any) => {
+    const cookie = (context as HocusPocusServerContext).cookie;
+    // query params
+    const params = requestParameters;
+    const documentType = params.get("documentType")?.toString() as TDocumentTypes | undefined;
+
+    // TODO: Fix this lint error.
+    // eslint-disable-next-line no-async-promise-executor
+    return new Promise(async () => {
```

Comment:
Does this logic need to execute later or is it fine to `awaitPageDescription()` etc immediately in this tick?